testnet/stacks-node/src/burnchains/bitcoin_regtest_controller.rs

```diff
@@ -787,50 +917,169 @@ impl BitcoinRegtestController {
+        signer: &mut BurnchainOpSigner,
+        _attempt: u64,
+    ) -> Option<Transaction> {
+        // Are we currently tracking an operation?
+        if self.ongoing_block_commit.is_none() {
+            // Good to go, let's build the transaction and send it.
+            let res = self.send_block_commit_operation(payload, signer, None, None, None, &vec![]);
+            return res;
+        }
+
+        let ongoing_op = self.ongoing_block_commit.take().unwrap();
+
+        let _ = self.sortdb_mut();
+        let burnchain_db = self.burnchain_db.as_ref().expect("BurnchainDB not opened");
+
+        for txid in ongoing_op.txids.iter() {
+            let mined_op = burnchain_db.get_burnchain_op(txid);
+            if mined_op.is_some() {
+                // Good to go, the transaction in progress was mined
+                debug!("Was able to retrieve ongoing TXID - {}", txid);
+                let res =
+                    self.send_block_commit_operation(payload, signer, None, None, None, &vec![]);
+                return res;
+            } else {
+                debug!("Was unable to retrieve ongoing TXID - {}", txid);
+            }
+        }
+
+        // Stop as soon as the fee_rate is 1.25 higher, stop RBF
+        if ongoing_op.fees.fee_rate > (self.config.burnchain.satoshis_per_byte * 150 / 100) {
+            let res = self.send_block_commit_operation(payload, signer, None, None, None, &vec![]);
+            return res;
+        }
+
+        // Did a re-org occurred since we fetched our UTXOs
+        if let Err(e) = burnchain_db.get_burnchain_block(&ongoing_op.utxos.bhh) {
+            info!(
+                "Possible presence of fork, invalidating cached set of UTXOs - {:?}",
+                e
+            );
+            let res = self.send_block_commit_operation(payload, signer, None, None, None, &vec![]);
+            return res;
+        }
+
+        // An ongoing operation is in the mempool and we received a new block. The desired behaviour is the following:
+        // 1) If the ongoing and the incoming operation are **strictly** identical, we will be idempotent and discard the incoming.
+        // 2) If the 2 operations are different, we will try to avoid wasting UTXOs, and attempt to RBF the outgoing transaction:
+        //  i) If UTXOs are insufficient,
+        //    a) If no other UTXOs, we'll have to wait on the ongoing operation to be mined before resuming operation.
+        //    b) If we have some other UTXOs, drop the ongoing operation, and track the new one.
+        //  ii) If UTXOs initially used are sufficient for paying for a fee bump, then RBF
+
+        // Let's start by early returning 1)
+        if payload == ongoing_op.payload {
+            info!("Abort attempt to re-submit identical LeaderBlockCommit");
+            return None;
+        }
+
+        // Let's proceed and early return 2) i)
+        let res = if ongoing_op.fees.estimated_amount_required() > ongoing_op.sum_utxos() {
```

Comment:
I think the code should try to avoid this case if at all possible.  Sending two block-commits ought to be strictly more expensive than sending one, and only one of these two block-commits is going to win sortition.  Maybe the code should instead attempt to figure out the total set of UTXOs to use in advance, given a maximum transaction fee, and then only try to RBF?

Reply:
Also, I think the code should stop attempting to send block-commits altogether if the Bitcoin block height advances, since the block-commit will almost certainly become invalid.  Is this handled anywhere?

Reply:
Not sure to understand the case you're challenging. `2) i)`?
> Sending two block-commits ought to be strictly more expensive than sending one, and only one of these two block-commits is going to win sortition.

If we're entering this codepath, it means that you have an ongoing transaction that was not mined in time, and the UTXOs used are too small for allowing a replacement by fee.
So the miner will indeed end up with 2 block commits in a block, but one of them will be outdated and won't be able to win the sortition.

Reply:
> it means that you have an ongoing transaction that was not mined in time

If this transaction has not been mined in time -- i.e. the Bitcoin blockchain advanced by one block -- then there's nothing more that this code can do.  The node needs to go and generate another Stacks block off of the new chain tip and re-try its bock-commit.

I'm saying that we should try to avoid the case where the transaction can no longer be RBF'ed because it's out of UTXOs to spend.  I think the node should instead have an "RBF budget" whereby it calculates the set of UTXOs required to pay for not only the initial transaction (with a typical fee), but also some additional number of RBFs that may be attempted in this Bitcoin block.  This way, the transaction would start out with enough UTXOs to pay for all of its RBF attempts up-front, and the node would simply decrement and re-sign the transaction's change address on each RBF.  If the Bitcoin block takes an unusually long time to be mined, and the node's RBF budget gets exceeded, then it should just give up attempting to win the block race for this block.  If it instead tried to send a second block-commit, it's forcing the miner to basically pay double for the chance to win the same block (since at least one of its in-flight block-commits is _guaranteed_ to fail).